nmrfx-processor/src/main/java/org/nmrfx/processor/datasets/vendor/JCAMPData.java

```diff
@@ -17,987 +17,569 @@
+        double offsetHz = block.optional(offsetLabel).map(JCampRecord::getDouble).orElse(0d);
+        return offsetHz / getSF(dim);
     }
 
     @Override
-    public int getSize(int iDim) {
-        return tdsize[iDim];
+    public void setRef(int dim, double value) {
+        ref.put(dim, value);
     }
 
     @Override
-    public int getMaxSize(int iDim) {
-        return maxSize[iDim];
+    public void resetRef(int dim) {
+       ref.remove(dim);
     }
 
     @Override
-    public void setSize(int iDim, int size) {
-        if (size > maxSize[iDim]) {
-            size = maxSize[iDim];
+    public double getRefPoint(int dim) {
+        // reference defined by getRef() is for the center of spectra
+        return getSize(dim) / 2.0;
+    }
+
+    @Override
+    public String getTN(int dim) {
+        if (dim == 0) {
+            return block.optional(_OBSERVE_NUCLEUS, $NUC_1, $T2_NUCLEUS)
+                    .map(JCampRecord::getString)
+                    .map(JCampUtil::toNucleusName)
+                    .orElse(NMRDataUtil.guessNucleusFromFreq(getSF(dim)).toString());
+        } else if (dim == 1) {
+            return block.optional($NUC_2)
+                    .map(JCampRecord::getString)
+                    .map(JCampUtil::toNucleusName)
+                    .orElse(NMRDataUtil.guessNucleusFromFreq(getSF(dim)).toString());
+        } else {
+            throw new UnsupportedOperationException("Unsupported dimension " + dim + " in JCamp");
         }
-        tdsize[iDim] = size;
     }
 
     @Override
-    public void setRef(int iDim, double ref) {
-        Ref[iDim] = ref;
-    }
+    public boolean isComplex(int dim) {
+        // For first dimension, check if the jcamp block contains imaginary pages
+        if (dim == 0) {
+            return !imaginaryPages.isEmpty();
+        }
 
-    @Override
-    public void resetRef(int iDim) {
-        Ref[iDim] = null;
-    }
+        // For other dimensions, infer it from FnMODE
+        FnMode fnMode = getFnMode(dim);
+        if (fnMode == FnMode.QSEQ || fnMode == FnMode.TPPI)
+            return false;
+        if (fnMode == FnMode.QF)
+            return getValues(dim).isEmpty();
 
-    @Override
-    public double getRef(int iDim) {
-        double ref = 0.0;
-        if (Ref[iDim] != null) {
-            ref = Ref[iDim];
-        } else {
-            Double dpar;
-            if ((dpar = getParDouble("TRANSMITPOS," + (iDim + 1))) != null) {
-                ref = dpar;
-            }
-            setRef(iDim, ref);
-        }
-        return ref;
+        return true;
     }
 
     @Override
-    public double getRefPoint(int dim) {
-        return 1.0;
-    }
+    public String getFTType(int dim) {
+        // known values: ft, rft (real), negate (hypercomplex)
+        //XXX original JCamp has "ft" hardcoded.
+        // Bruker is using AQ_Mod and FnMode, which is what I chose to copy here.
+        // Not whether it should be filled for FID as well.
```

Comment:
Not whether it should be filled for FID as well. Would be better with an enum instead of a string, but this is out of scope here.

Reply:
This is used in setting up processing scripts in python code so while an enum would be better in Java, the string was certainly simpler in Python (probably a way to use enums in Jython, but I"m not sure I've done it).